lib/backend/etcd/etcd.go

```diff
@@ -300,3 +326,79 @@ func convertEtcdError(err error, key model.Key) error {
 		return errors.ErrorDatastoreError{Err: err, Identifier: key}
 	}
 }
+
+// getHostMetadataFromDirectory gets hosts that may not be configured with a host
+// metadata (older deployments or Openstack deployments).
+func (c *EtcdClient) getHostMetadataFromDirectory(k model.Key) (*model.KVPair, error) {
+	// The delete path of the host metadata includes the whole of the per-host
+	// felix tree, so check the existence of this tree and return and empty
+	// Metadata if it exists.
+	key, err := model.KeyToDefaultDeletePath(k)
+	if err != nil {
+		return nil, err
+	}
+	if _, err := c.etcdKeysAPI.Get(context.Background(), key, etcdGetOpts); err != nil {
+		return nil, convertEtcdError(err, k)
+	}
+
+	// The node exists, so return an empty Metadata.
+	kv := &model.KVPair{
+		Key:   k,
+		Value: &model.HostMetadata{},
+	}
+	return kv, nil
+}
+
+func (c *EtcdClient) listHostMetadata(l model.HostMetadataListOptions) ([]*model.KVPair, error) {
+	// If the hostname is specified then just attempt to get the host,
+	// returning an empty string if it does not exist.
+	if l.Hostname != "" {
+		log.Info("Listing host metadata with exact key")
+		hmk := model.HostMetadataKey{
+			Hostname: l.Hostname,
+		}
+		if kv, err := c.Get(hmk); err == nil {
+			return []*model.KVPair{kv}, nil
+		} else {
+			err = convertEtcdError(err, nil)
+			switch err.(type) {
+			case errors.ErrorResourceDoesNotExist:
+				return []*model.KVPair{}, nil
+			default:
+				return nil, err
+			}
+		}
+	}
+
+	// No hostname specified, so enumerate the directories directly under
+	// the host tree, return no entries if the host directory does not exist.
+	log.Info("Listing all host metadatas")
+	key := "/calico/v1/host"
+	if results, err := c.etcdKeysAPI.Get(context.Background(), key, etcdListChildrenOpts); err != nil {
+		// If the root key does not exist - that's fine, return no list entries.
+		log.Info("Error enumerating host directories")
+		err = convertEtcdError(err, nil)
+		switch err.(type) {
+		case errors.ErrorResourceDoesNotExist:
+			return []*model.KVPair{}, nil
+		default:
+			return nil, err
+		}
+	} else {
+		// TODO:  Since the host metadata is currently empty, we don't need
+		// to perform an additional get here, but in the future when the metadata
+		// may contain fields, we would need to perform a get.
+		log.Info("Parse host directories.")
```

Comment:
I'd probably use `Debug` to avoid spam if something like Felix does a read, WDYT?